File: Android/src/androidMain/kotlin/com/taskbridge/android/repository/impl/MessagesRepositoryImpl.kt
package com.taskbridge.android.repository.impl

import com.taskbridge.android.repository.MessagesRepository
import com.taskbridge.core.interactors.messages.MessagesInteractor
import com.taskbridge.core.models.messages.AppMessage
import com.taskbridge.core.models.messages.AppMessageKey
import kotlinx.coroutines.flow.Flow

class MessagesRepositoryImpl(
    private val interactor: MessagesInteractor
) : MessagesRepository {
    override fun observeAll(): Flow<AppMessage> {
        return interactor.observeAll()
    }

    override fun observe(types: List<AppMessageKey>): Flow<AppMessage> {
        return interactor.observe(types)
    }

    override fun observe(type: AppMessageKey): Flow<AppMessage> {
        return interactor.observe(type)
    }

    override suspend fun publish(message: AppMessage) {
        interactor.publish(message = message)
    }
}

File: iOS/TaskBridge/Repository/Impl/MessagesRepositoryImpl.swift
import Foundation
import TaskBridgeCore

@MainActor
class MessagesRepositoryImpl: MessagesRepository {
    private let interactor: MessagesInteractor

    init(interactor: MessagesInteractor) {
        self.interactor = interactor
    }

    func observeAll() -> AsyncStream<AppMessage> {
        return createStream(for: interactor.observeAll())
    }

    func observe(types: [AppMessageKey]) -> AsyncStream<AppMessage> {
        return createStream(for: interactor.observe(types: types))
    }

    func observe(type: AppMessageKey) -> AsyncStream<AppMessage> {
        return createStream(for: interactor.observe(type: type))
    }

    func publish(message: AppMessage) async throws {
        try await interactor.publish(message: message)
    }
}
